### DISCUSSION TITLE
Shading and Shadow questions

### DISCUSSION THREAD

**Discussion by aVersionOfReality** [bnpr/Malt#127] [Q&A]
Hello! Some questions about shadows and shaders.

1) I want to get shadows isolated on their own. What is the best way to do this? I have gotten them so far by using the scene_toon node and setting the offset to 1.
![image](https://user-images.githubusercontent.com/36803956/133582382-94b9e350-61c6-449b-bf4c-dd86c6929389.png)

2) In the above picture, we can see jagged lines at the edges of the shadows. What can be done about this? I have seen it improve by turning up the resolution on Sun Cascades. But even at 4096, the issue is visible. Should I be needing to have it that high or higher?

3) How are these settings used?
![image](https://user-images.githubusercontent.com/36803956/133585048-5319e5ee-2d0f-41cf-b0ea-1dcff31ef070.png)

4) I want to have a material that receives shadows cast by other objects, but does not cast any shadows of its own. In Cycles/Eevee, you can control the strength of the shadows something casts by using the Light Path node's Is Shadow Ray to mix between the surface shader and a transparency shader. Is there any equivalent to the Light Path node's outputs, or any other way to do this?

5) What exactly is the difference between scene_toon, get_toon, and lit_toon? (or any of the others.) It seems that scene versions have more input options?

6) I want to use custom normals with AO and Curvature. There does not seem to be any scene versions of this. Is this possible in some other way? I see the ao_ex and Curvature nodes in the Filters section seem to have options for normals to be inputted, but I cannot figure out how to use these. They seem to expect image texture inputs?

Thank you!

**Comment by pragma37** [comment]
1. By setting the Toon Offset to 1, you'll only get the shadows coming from the shadow maps. This might sound like what you want, but you are going to get shadow termination issues. It's better to set the toon size to 0.49 or less and everything else to 0.
I personally prefer to use the diffuse gradient node.

2. If you need more shadow resolution you can always increase the resolution setting and the number of cascades, and decrease the max distance to the minimum possible. I think 4096 is reasonable for final rendering.
I have some code that improves shadow filtering, so it works better at lower resolutions, but it's not yet ready for release.

3. The global Light Group settings are the ones that will be used when you use the "basic" shading functions (get_diffuse, get_diffuse_gradient, get_specular...) from the NPR Mesh Shader section.

So, with Light (1, 0, 0, 0) the material will only be affected by lights if its Light Group is equal to 1 or 0.
By default, the Light Group for new lights is 1.
If you have, for example, (1, 5, 0, 0), it will be lit by Light Groups 1, 5 and 0, but it won't be lit, for example, by Light Groups 2, 3, 4, 6... etc.

The Shadow Channels is the same, but for casting shadows.
So a material could cast shadows on a Light Group, but don't get lit by it. Or get lit by it, but don't cast shadows.

4. Assuming all your Lights are in Light Group 1, just set the material Light Group > Shadow to (0, 0, 0, 0).

Additionally, I just added new functions to check in what pass is being rendered: 9a2fa59e83644e1a03ba401ac0974d472e6d8a4e

5. I think this is best understood by looking at the source. Basically:
- get_diffuse calls scene_diffuse for each Light Group in the material settings and with the default position and normal.
https://github.com/bnpr/Malt/blob/9a2fa59e83644e1a03ba401ac0974d472e6d8a4e/Malt/Shaders/Pipelines/NPR_Pipeline/NPR_MeshShader.glsl#L280
- scene_diffuse calls lit_diffuse for each light that matches the Light Group parameter.
https://github.com/bnpr/Malt/blob/9a2fa59e83644e1a03ba401ac0974d472e6d8a4e/Malt/Shaders/Pipelines/NPR_Pipeline/NPR_Shading.glsl#L108
- And lit_diffuse calculates the light contribution for a single light.
https://github.com/bnpr/Malt/blob/9a2fa59e83644e1a03ba401ac0974d472e6d8a4e/Malt/Shaders/Pipelines/NPR_Pipeline/NPR_Shading.glsl#L9

6. AO and Curvature use the depth and normals from the PRE_PASS. So they use whatever normal you set in the material's PixelOutput.
In the default NPR Pipeline, materials are rendered in 2 steps:

1º.  PRE_PASS. It has 2 Render Targets (AOVs if you prefer), a World Space Normal (RGB) and Depth (Alpha) texture, and an ID texture with just 1 channel.

These textures are then set as uniform parameters in the MAIN_PASS.

2º. MAIN_PASS. It has 3 Render Targets. Render Color, Line Color, and Line Data (Width).

https://github.com/bnpr/Malt/blob/9a2fa59e83644e1a03ba401ac0974d472e6d8a4e/Malt/Shaders/Pipelines/NPR_Pipeline/NPR_MeshShader.glsl#L142

So any shader that requires info from nearby pixels (AO, bevel, curvature, line detection...) is going to read from those textures.
So the difference between get_curvature and curvature is kind of the same as with get_diffuse and scene_diffuse.
get_curvatures is just a wrapper around curvature:
https://github.com/bnpr/Malt/blob/9a2fa59e83644e1a03ba401ac0974d472e6d8a4e/Malt/Shaders/Pipelines/NPR_Pipeline/NPR_MeshShader.glsl#L377

**Reply by aVersionOfReality** [reply]
1-4: Okay, got all that working. As I was testing the diffuse gradient node, I noticed that on the node tree I was working in, nodes cannot displace the gradients properly. I can select one, but cannot create or edit them. However, if I switch to a new node tree, it works fine. Is this a bug, or is there some way I have somehow disabled displaying those controls for just one node tree?

5: Great, and I figured out how to actually use the lit surfaces too. So I can basically create a light wholly in the material.

6: I was trying to bevel the AO which wasn't working, hence the original question. Now I'm trying it again and it _is_ working with curvature (using BevelSettings version) but still does not work with AO. I'm not sure why. Is there something different about AO? Beveling it works in Cycles.

How exactly do I access the stuff from the pre-pass? I see the various filter nodes take normal and/or depth textures as inputs, but the node UI only lets me select image textures for these. I cannot find any node that outputs the right socket type. Is there one? If not, could it be made by just making a utility that outputs the correct property?

And how does this factor into the different methods of beveling? I understand that get_bevel will be beveling the normals that are inputted into the PixelOutput's normal socket. So if I plugged this back into the PixelOutput, that is a circular loop, right? (I've tried it and it does nothing.) Can we not use bevel_ex to bevel the normals from the pre-pass or something? (I assume this is what bevel_settings is doing, but I'm not clear on why it needs to be done the way it is. Just trying to fill in the rest of my mental image of how this all fits together.)

Thanks!

**Comment by pragma37** [comment]
1-4. Sounds like a bug. Could you upload a blend file with the material where the gradients are locked?

6. AO doesn't use take normals into account, just depth. Cycles is a ray-tracer, so, completely different worlds.

Using pass-specific textures from nodes would be tricky. It's better to just use the NPR MeshShader nodes, that's why they are there.
If you want to test the filter nodes directly, use them from screen-shaders.

>  I understand that get_bevel will be beveling the normals that are inputted into the PixelOutput's normal socket. So if I plugged this back into the PixelOutput, that is a circular loop, right?

get_bevel will bevel **in the MAIN PASS**, the normals that where output **in the PRE PASS**.

> Can we not use bevel_ex to bevel the normals from the pre-pass or something?

That's exactly what get_bevel does.
